File: Chain_of_Responsibility_Pattern/go/Chain_of_Responsibility.go
package Chain_of_Responsibility

import "fmt"

type Logger interface {
	write(message string)
	SetNextLog(Logger)
	Next() Logger
	Level() int
}

type Log struct {
	LoggerChain Logger
}

func (this *Log) Log(level int, message string) {
	c := this.LoggerChain
	for c != nil {
		if level != c.Level() {
			continue
		}
		c.write(message)
		c = c.Next()
	}
}

type Debug struct {
	level      int
	nextLogger Logger
}

func (this *Debug) write(message string) {
	fmt.Printf("[DEBUG] %s\n", message)
}
func (this *Debug) SetNextLog(l Logger) {
	this.nextLogger = l
}
func (this *Debug) Level() int {
	return this.level
}
func (this *Debug) Next() Logger {
	return this.nextLogger
}

type Info struct {
	level      int
	nextLogger Logger
}

func (this *Info) write(message string) {
	fmt.Printf("[INFO] %s\n", message)
}
func (this *Info) SetNextLog(l Logger) {
	this.nextLogger = l
}
func (this *Info) Level() int {
	return this.level
}
func (this *Info) Next() Logger {
	return this.nextLogger
}

type Error struct {
	level      int
	nextLogger Logger
}

func (this *Error) write(message string) {
	fmt.Printf("[ERROR] %s\n", message)
}
func (this *Error) SetNextLog(l Logger) {
	this.nextLogger = l
}
func (this *Error) Level() int {
	return this.level
}
func (this *Error) Next() Logger {
	return this.nextLogger
}
func GetLogger() *Log {
	eLog := Error{level: 1}
	iLog := Info{level: 2}
	dLog := Debug{level: 3}

	dLog.SetNextLog(&iLog)
	iLog.SetNextLog(&eLog)

	l := Log{LoggerChain: &eLog}

	return &l
}

File: Chain_of_Responsibility_Pattern/cpp/Chain_of_responsibility.cc
#include <iostream>
using namespace std;
class Logger {
public:
    int level;
    Logger* nextLogger;
public:
    virtual void Log(int leve, const string& message) {
        if (leve == level) {
            write(message);
        } else {
            if (nextLogger) {
                nextLogger->Log(leve, message);
            } else {
                cout << "unknow level " << leve <<endl;
            }
        }
    }
    virtual void setNextLog(Logger* next) {
        nextLogger = next;
    }
private:
    virtual void write(const string& message) = 0;
};

class Debug : public Logger {
private:
    void write(const string& message) {
        cout << "[DEBUG] " << message << endl;
    } 
};

class Info : public Logger {
private:
    void write(const string& message) {
        cout << "[Info] " << message << endl;
    } 
};
class Error : public Logger {
private:
    void write(const string& message) {
        cout << "[ERROR] " << message << endl;
    } 
};

Logger* GetLogger() {
    Debug* d = new Debug;
    d->level = 1;

    Info* i = new Info;
    i->level = 2;

    Error* e = new Error;
    e->level = 3;

    d->setNextLog(i);
    i->setNextLog(e);

    return d;
}

void test_chainof() {
    Logger* log = GetLogger();
    log->Log(1, "hello");
    log->Log(4, "hello2");
}
int main() {
    test_chainof();
    return 0;
}